sharder/deterministic.go

```diff
@@ -117,33 +118,54 @@ func (d *DeterministicSharder) Start() error {
 		}
 		d.Logger.Debug().Logf("picked up local peer port of %s", localPort)
 
-		// get my local interfaces
+		// get my local interfaces' IPs.
 		localAddrs, err := net.InterfaceAddrs()
 		if err != nil {
 			return errors.Wrap(err, "failed to get local interface list to initialize sharder")
 		}
+		localIPs := make([]string, len(localAddrs))
+		for i, addr := range localAddrs {
+			addrStr := addr.String()
+			ip, _, err := net.ParseCIDR(addrStr)
+			if err != nil {
+				return errors.Wrap(err, fmt.Sprintf("failed to parse CIDR for local IP %s", addrStr))
+			}
+			localIPs[i] = ip.String()
+		}
+
+		// If RedisIdentifier is an IP, add it to localIps.
+		if redisIdentifier, err := d.Config.GetRedisIdentifier(); err == nil && redisIdentifier != "" {
+			if ip := net.ParseIP(redisIdentifier); ip != nil {
+				d.Logger.Debug().Logf("Using RedisIdentifier as public IP: %s", redisIdentifier)
+				localIPs = append(localIPs, redisIdentifier)
+			}
+		}
 
 		// go through peer list, resolve each address, see if any of them match any
 		// local interface. Note that this assumes only one instance of Refinery per
 		// host can run.
 		for i, peerShard := range d.peers {
-			d.Logger.Debug().WithField("peer", peerShard).WithField("self", localAddrs).Logf("Considering peer looking for self")
+			d.Logger.Debug().WithFields(logrus.Fields{
+				"peer": peerShard,
+				"self": localIPs,
+			}).Logf("Considering peer looking for self")
 			peerIPList, err := net.LookupHost(peerShard.ipOrHost)
 			if err != nil {
 				// TODO something better than fail to start if peer is missing
 				return errors.Wrap(err, fmt.Sprintf("couldn't resolve peer hostname %s", peerShard.ipOrHost))
 			}
 			for _, peerIP := range peerIPList {
-				for _, localIP := range localAddrs {
-					ipAddr, _, err := net.ParseCIDR(localIP.String())
-					if err != nil {
-						return errors.Wrap(err, fmt.Sprintf("failed to parse CIDR for local IP %s", localIP.String()))
-					}
-					if peerIP == ipAddr.String() {
+				for _, ipAddr := range localIPs {
+					if peerIP == ipAddr {
 						if peerShard.port == localPort {
 							d.Logger.Debug().WithField("peer", peerShard).Logf("Found myself in peer list")
 							found = true
 							selfIndexIntoPeerList = i
+						} else {
```

Comment:
Did you find a case where this happens? I'm curious as to why you added this.

Reply:
Yes, and it was ultimately a misconfiguration.  The `RedisIdentifier` port and `PeerListenAddr` port must be identical.  This required configuring static ports in Nomad, where randomly generated port mappings are the default.